### PR TITLE
[NO SQUASH] ItemDefinition: Add wield_offset and wield_rotation

### LINKED ISSUE #7762: Ability to specify wielditem rotation and wielditem offset in all three axes
##### Issue type

 - Feature request

##### Summary

Currently, wielditems can only be rotated by using the texture modifier `[transform`. It'd be much more useful to have a "fully-featured" field for wield_rotation.

I've also felt the need to offset the wielditem in a certain direction many a times. This is _also_ due to the fact that a small portion of the wielditem is always clipped due to its position. But even so, it would be very helpful to be able to specify offset for wielditems, similar to the `offset` field in a HUD element def.

##### Example implementation
Note that this is just an example for the new itemdef fields, and `x`, `y`, `z` are dummies. `wield_rotation` could use radians, but not sure about what `wield_offset` would use...
```lua
minetest.register_item("mymod:name", {
	...,
	wield_rotation = {x = 15, y = 30, z = 45},
	wield_offset = {x = 10, y = 20, z = 30},
	...
})
```

### PR DESCRIPTION
**Pull request by ClobberXD**
This PR adds two new fields to the item definition table:

- `wield_rotation = { x = , y = , z = }` - Absolute rotation of wieldmesh in degrees. If not specified, a default `v3f` of `[-100,120,-100]` is assumed, for the sake of backwards compat.

- `wield_position = { x = , y = }` - Absolute position of wieldmesh on screen, with values supported from 0 to 1 for both `x` and `y` (`[0,0]` == bottom-left corner, `[1,1]` == top-right). If not specified, a default v2f `[55,-35]` (non-normalized values) is assumed, for the sake of backwards-compat.

  The reason this is a 2D vector is because moving this along the Z-axis doesn't make much sense. If required, one could make the wielditem smaller by using `wield_scale` instead of offsetting it in the +Z axis.

These transformations are applied every time `WieldMeshSceneNode::setItem` is called.

### Use-cases

- This feature is primarily intended to go hand-in-hand with #8725. A classic mesh item use-case would be guns - they need to be rotated to point exactly straight ahead, and they also need to be moved to the centre for aiming down sights.
- Different orientations for different tools like pickaxes, shovels, axes, swords, etc.

### How to test

--- TO BE UPDATED ---

****

This PR is still WIP. Closes #7762.

### CODE REVIEW

**Comment by DS-Minetest** [comment]
In your test mod, `core.override_item` is used `on_use` and `on_secondary_use`. This is horrible. It changes the rotation and offset for all items of that type and, according to `lua_api.txt`, `core.override_item` should only be called on load time. If offset and rotation should be changeable in runtime, item meta has to be used.

What would actual usecases of a static wielditem rotation and offset be?

**Comment by ClobberXD** [comment]
> `core.override_item` should only be called on load time

Ah, lol. That could be one reason why this PR _apparently_ doesn't work. :P 

I used `override_item` in my test mod only for demonstration purposes. These fields are supposed to be defined at load-time and aren't created with the intention of runtime modification.

> What would actual usecases of a static wielditem rotation and offset be?

I've updated the first post with a bunch of use-cases from the top of my head.

**Comment by DS-Minetest** [comment]
:thinking: The usecases you came up with could also be solved with per player wieldhand offset and rotation.

(A bit related: #2811)

**Comment by ClobberXD** [comment]
Well yes, but actually no. You see, while per-player offset would technically solve the problem too, there's no need to make these fields per-player, when they only need to differ per-item.

**Comment by ClobberXD** [comment]
I've split the code-style fixes into separate commits; this PR is now a `no-squash` PR.

**Comment by ClobberXD** [comment]
As mentioned in the previous post, the core API functionality is already in place, and the only stuff remaining (AFAICT) are:

- All items currently use a hard-coded offset to render them to the right side of the screen. Should the wield offset  be based on this hard-coded offset, or should it be based on 0,0 (centre of the screen)? Regarding documentation,
- I'm unable to figure out the units of `wield_offset`, which seems to be using some arbitrary values. This is expected, as render offsets can't really be based on units like BS or BS * 10, but makes documentation a bit of a hassle for me. :)

Inputs appreciated.

**Comment by LoneWolfHT** [comment]
I'm willing to test this when it's ready

**Comment by ClobberXD** [comment]
Sweet! I'll be finishing #8725 first (probably by this evening), and then work on this.

**Comment by ClobberXD** [comment]
Here's another use-case for `wield_rotation` in case anyone isn't fully convinced about the usefulness of this feature - arbitrary rotation without encountering problems with `[transform` as detailed in #8465.

**Comment by ClobberXD** [comment]
Hello, core dev inputs needed on https://github.com/minetest/minetest/pull/8755#issuecomment-573262510. General feedback would be appreciated as well. Once this is resolved, the PR would be ready for review.

File: src/client/wieldmesh.cpp [hunk 1]
```diff
@@ -580,7 +576,11 @@ void getItemMesh(Client *client, const ItemStack &item, ItemMesh *result)
 
 		rotateMeshXZby(mesh, -45);
 		rotateMeshYZby(mesh, -30);
+	} else {
+		result->mesh = mesh;
+		return;
```
Comment (SmallJoker) [inline src/client/wieldmesh.cpp:581]: Redundant code.

File: src/client/wieldmesh.cpp [hunk 2]
```diff
@@ -357,8 +357,7 @@ void WieldMeshSceneNode::setItem(const ItemStack &item, Client *client, bool che
 
 	// If wield_image needs to be checked and is defined, it overrides everything else
 	if (!def.wield_image.empty() && check_wield_image) {
-		setExtruded(def.wield_image, def.wield_overlay, def.wield_scale, tsrc,
-			1);
+		setExtruded(def.wield_image, def.wield_overlay, def.wield_scale, tsrc, 1);
```
Comment (SmallJoker) [inline src/client/wieldmesh.cpp:360]: No rotation for extruded meshes?

File: src/client/wieldmesh.cpp [hunk 3]
```diff
@@ -432,19 +430,17 @@ void WieldMeshSceneNode::setItem(const ItemStack &item, Client *client, bool che
...
 		setExtruded(def.inventory_image, def.inventory_overlay, def.wield_scale,
 			tsrc, 1);
 		m_colors.emplace_back();
 		// overlay is white, if present
 		m_colors.emplace_back(true, video::SColor(0xFFFFFFFF));
-		return;
 	}
 
-	// no wield mesh found
-	changeToMesh(nullptr);
```
Comment (SmallJoker) [inline src/client/wieldmesh.cpp:447]: This call should be kept to properly update the mesh when none of the cases above matches.
```C++
	m_colors.emplace_back(true, video::SColor(0xFFFFFFFF));
} else {
	// no wield mesh found
	changeToMesh(nullptr);
}
```

EDIT: Only works if all cases use the same `else if` chain.

**Comment by ClobberXD** [comment]
@SmallJoker Code-style commits are in their own separate commits. The actual feature commit is the first commit in this PR, followed by the code-style commits.

File: src/client/camera.cpp [hunk 4]
```diff
@@ -580,36 +569,33 @@ void Camera::setDigging(s32 button)
 
 void Camera::wield(const ItemStack &item)
 {
-	if (item.name != m_wield_item_next.name ||
-			item.metadata != m_wield_item_next.metadata) {
+	if (item != m_wield_item_next) {
```
[the remark this answers is not in the document]
Comment (ClobberXD) [reply src/client/camera.cpp:590]: You're right, I'll revert this.

File: src/client/wieldmesh.cpp (hunk 3 above)
Comment (ClobberXD) [reply src/client/wieldmesh.cpp:447]: Yessir, gotit.

File: src/client/wieldmesh.cpp (hunk 1 above)
Comment (ClobberXD) [reply src/client/wieldmesh.cpp:596]: True, I don't remember why I added this.

**Comment by ClobberXD** [comment]
I realised that the original feature commit contained code-style changes as well. I apologise for the confusion it caused. I've now re-consolidated the changes, and the feature commit contains only the feature-related code now.

****

I've rebased this PR, but I haven't attended to all the review comments. I'm planning to also overhaul this feature, as it doesn't make sense to start from arbitrary an position/rotation respectively:

- I'm planning to convert the field into `wield_pos` (a `v2f` again), which would support values from 0 to 1 (normalised coords) - [0,0] would be the top-left corner of the screen, and [1,1] would be the bottom right.
  Let me know if the lower bound should be -1 instead.
- If `wield_pos` isn't explicitly defined, the wieldmeshnode will be placed in the default arbitrary location to preserve backwards compat. The same goes for `wield_rotation` - if not explicitly defined, the wieldmeshnode will be rotated in the arbitrary rotation that's currently being used.

**Pull request closed without merging**